aggregator/follower_flush_mgr.go

```diff
@@ -0,0 +1,279 @@
+	watchCreateErrors  tally.Counter
+	unmarshalErrors    tally.Counter
+	shardNotFound      tally.Counter
+	resolutionNotFound tally.Counter
+	kvUpdateFlush      tally.Counter
+	forcedFlush        tally.Counter
+}
+
+func newFollowerFlushManagerMetrics(scope tally.Scope) followerFlushManagerMetrics {
+	return followerFlushManagerMetrics{
+		watchCreateErrors:  scope.Counter("watch-create-errors"),
+		unmarshalErrors:    scope.Counter("unmarshal-errors"),
+		shardNotFound:      scope.Counter("shard-not-found"),
+		resolutionNotFound: scope.Counter("resolution-not-found"),
+		kvUpdateFlush:      scope.Counter("kv-update-flush"),
+		forcedFlush:        scope.Counter("forced-flush"),
+	}
+}
+
+type followerFlushManager struct {
+	sync.RWMutex
+
+	nowFn                 clock.NowFn
+	checkEvery            time.Duration
+	workers               xsync.WorkerPool
+	flushTimesKeyFmt      string
+	flushTimesStore       kv.Store
+	maxNoFlushDuration    time.Duration
+	forcedFlushWindowSize time.Duration
+	logger                xlog.Logger
+	scope                 tally.Scope
+
+	doneCh            <-chan struct{}
+	proto             schema.ShardSetFlushTimes
+	flushTimesKey     string
+	flushTimesUpdated bool
+	lastFlushed       time.Time
+	sleepFn           sleepFn
+	metrics           followerFlushManagerMetrics
+}
+
+func newFollowerFlushManager(
+	doneCh <-chan struct{},
+	opts FlushManagerOptions,
+) roleBasedFlushManager {
+	nowFn := opts.ClockOptions().NowFn()
+	instrumentOpts := opts.InstrumentOptions()
+	scope := instrumentOpts.MetricsScope()
+	return &followerFlushManager{
+		nowFn:                 nowFn,
+		checkEvery:            opts.CheckEvery(),
+		workers:               opts.WorkerPool(),
+		flushTimesKeyFmt:      opts.FlushTimesKeyFmt(),
+		flushTimesStore:       opts.FlushTimesStore(),
+		maxNoFlushDuration:    opts.MaxNoFlushDuration(),
+		forcedFlushWindowSize: opts.ForcedFlushWindowSize(),
+		logger:                instrumentOpts.Logger(),
+		scope:                 scope,
+		doneCh:                doneCh,
+		lastFlushed:           nowFn(),
+		sleepFn:               time.Sleep,
+		metrics:               newFollowerFlushManagerMetrics(scope),
+	}
+}
+
+func (mgr *followerFlushManager) Open(shardSetID string) {
+	mgr.Lock()
+	defer mgr.Unlock()
+
+	mgr.flushTimesKey = fmt.Sprintf(mgr.flushTimesKeyFmt, shardSetID)
+	go mgr.watchFlushTimes()
+}
+
+// NB(xichen): no actions needed for initializing the follower flush manager.
+func (mgr *followerFlushManager) Init(*sync.RWMutex, []*flushBucket) {}
+
+func (mgr *followerFlushManager) Flush(
+	bucketLock *sync.RWMutex,
+	buckets []*flushBucket,
+) (bool, time.Duration) {
+	var (
+		shouldFlush        = false
+		flushersByInterval []flushersGroup
+	)
+	// NB(xichen): a flush is triggered in the following scenarios:
+	// * The flush times persisted in kv have been updated since last flush.
+	// * Sufficient time (a.k.a. maxNoFlushDuration) has elapsed since last flush.
+	bucketLock.RLock()
```

Comment:
The `bucketLock` is used to synchronize access to `buckets`. The `buckets` are kept in the top-level flush manager as a global registry for the lists containing aggregated metrics in different shards, and are shared between the leader flush manager and the follower flush manager for centralized access to the registered lists. As such, the bucket lock also needs to be shared between the top-level flush manager, the leader flush manager, and the follower flush manager and cannot be merged with the individual manager locks.